File: Roles.py
import random
import string 

class Utilisateur: 
    def __init__(self, nom, prenom, email, password): 
        self.nom = nom
        self.prenom = prenom
        self.email = email
        self.password = password
        self.role = 'Sans role'
        self.estActive = True

class Quiz:
    def __init__(self, titre, typeDeQuiz, cycleDeVie, choix):
        self.titre = titre
        self.typeDeQuiz = typeDeQuiz
        self.cycleDevie = cycleDeVie
        self.choix = choix
        self.estActive = True

class Administrateur(Utilisateur):
    def __init__(self, admin : Utilisateur):
        super().__init__(admin)
        self.role = 'Administrateur'
        
    def creerUtilisateur(self, user : Utilisateur):
        return Utilisateur(user)
    
    def creerQuiz(self, quizz : Quiz):
        return Quiz(quizz)
    
    def desactiverUtilisateur(self, user : Utilisateur): 
        user.estActive = False
        print("L'utilisateur' {user.prenom} a été désactivé.")
    
    def activerUtilisateur(self, user : Utilisateur): 
        print("L'utilisateur {user.prenom} a été  activé.")
    
    def desactiverQuiz(self, quizz : Quiz): 
        quizz.estActive = False
        print("Le quiz {quizz.titre} a été désactivé.")
    
    def activerQuiz(self, quizz : Quiz): 
        print("Le quiz {quizz.titre} a été activé.")

class Validateur(Utilisateur):
    def __init__(self, validateur : Utilisateur):
        super().__init__(validateur)
        self.role = 'Validateur'

    def validerCompte(self, user : Utilisateur):
        user.estActive = True
        print("Le compte de l'utilisateur {user.prenom} {user.nom} a été validé.")

    def creerQuiz(self, quizz : Quiz):
        return Quiz(quizz)

class AdministrateurQuiz(Utilisateur):
    def __init__(self, adminQuiz : Utilisateur):
        super().__init__(adminQuiz)
        self.role = 'Administrateur de Quiz'
    
    def desactiverQuiz(self, quizz : Quiz): 
        quizz.estActive = False
        print("L'utilisateur {quizz.titre} a été  désactivé.")
        
    def creerQuiz(self, quizz : Quiz):
        return Quiz(quizz)

class Createur(Utilisateur):
    def __init__(self, createur : Utilisateur):
        super().__init__(createur)
        self.role = 'Createur'

    def creerQuiz(self, quizz : Quiz):
        return Quiz(quizz)

class UtilisateurParDefaut(Utilisateur):
    def __init__(self, userParDefaut : Utilisateur):
        super().__init__(userParDefaut)
        self.role = 'Utilisateur'

    def changerMotDePasse(self, newPassword, user:Utilisateur):
        user.password = newPassword
        print("Mot de passe changé avec succès.")

File: Roles.php
<?php

class Utilisateur {
    protected $nom;
    protected $prenom;
    protected $email;
    protected $password;
    protected $role;
    protected $estActive = TRUE;

    function __construct($nom, $prenom, $email, $password) {
        $this->nom = $nom;
        $this->prenom = $prenom;
        $this->email = $email;
        $this->password = $password;
        $this->role = 'Sans role';
        $this->estActive;
    }
}

class Quiz {
    protected $titre;
    protected $typeDeQuiz;
    protected $cycleDeVie;
    protected $choix;
    protected $questionQuizz=array();

    function __construct($titre, $typeDeQuiz, $cycleDeVie, $choix) {
        $this->titre = $titre;
        $this->typeDeQuiz = $typeDeQuiz;
        $this->cycleDeVie = $cycleDeVie;
        $this->choix = $choix;
        $this->questionQuizz = array();
    }

    function ajouter_question(Quiz $quizz, $questions) {
        if($quizz->cycleDeVie == 'Creation') {
            array_push($quizz->questionQuizz,$questions);
        } else {
            echo "Impossible d'ajouter des questions à un quizz Lancé ou Terminé";
        }
    }  

    function repondre_quizz(Utilisateur $user, Quiz $quizz, $reponses) {
        if($quizz->cycleDeVie == 'Lancé') {
            foreach ($quizz->questions as $question) {
                echo "Question : {$question->texte}\n";
                $reponseUtilisateur = readline("Votre réponse (Vrai/Faux) : ");
                if ($reponseUtilisateur === $question->reponseAttendue) {
                    echo "Bonne réponse !\n";
                } else {
                    echo "Mauvaise réponse !\n";
                }
            }
        } else {
            echo "Impossible de répondre au quiz car il n'est pas en cours.\n";
        }
    }

    function modifier_quiz(Quiz $quizz) {
        if ($quizz->cycleDeVie == 'Creation') {
            $this->titre = $quizz->titre;
            $this->typeDeQuiz = $quizz->typeDeQuiz;
            $this->choix = $quizz->choix;
        } else {
            echo "Impossible de modifier le quiz car il n'est pas en cours de création.";
        }
    }

    function verifier_etat_creation(Quiz $quizz) {
        if ($quizz->cycleDeVie == "Creation") {
            echo "Le quiz est actuellement en cours de création.";
        } elseif ($quizz->cycleDeVie == "Lance") {
            echo "Le quiz a été terminé et ne peut plus être modifié.";
        } else {
            echo "Le quiz est dans un état inconnu.";
        }
    }
}

class Administrateur extends Utilisateur {

    function __construct($nom, $prenom, $email, $password) {
        parent::__construct($nom, $prenom, $email, $password);
        $this->role = 'Administrateur'; 
    }

    function creerUtilisateur($nom, $prenom, $email, $password) {
        return new Utilisateur($nom, $prenom, $email, $password);
    }

    function creerQuiz($titre, $typeDeQuiz, $cycleDeVie, $choix) {
        return new Quiz($titre, $typeDeQuiz, $cycleDeVie, $choix);
    }

    function desactiverUtilisateur(Utilisateur $user) {
        $user->estActive = false;
        echo "L'utilisateur $user->prenom a été désactivé.";
    }

    function activerUtilisateur(Utilisateur $user) {
        echo "L'utilisateur $user->prenom a été activé.";
    }

    function desactiverQuiz(Quiz $quizz) {
        $quizz->estActive = false;
        echo "Le quiz $quizz->titre a été désactivé.";
    }

    function activerQuiz(Quiz $quizz) {
        echo "Le quiz $quizz->titre a été activé.";
    }
}

class Validateur extends Utilisateur {

    function __construct($nom, $prenom, $email, $password) {
        parent::__construct($nom, $prenom, $email, $password);
        $this->role = 'Validateur'; 
    }

    function validerCompte(Utilisateur $user) {
        $user->estActive = true;
        echo "Le compte de l'utilisateur $user->prenom $user->nom a été validé.";
    }

    function creerQuiz($titre, $typeDeQuiz, $cycleDeVie, $choix) {
        return new Quiz($titre, $typeDeQuiz, $cycleDeVie, $choix);
    }
}

class AdministrateurQuiz extends Utilisateur {

    function __construct($nom, $prenom, $email, $password) {
        parent::__construct($nom, $prenom, $email, $password);
        $this->role = 'Administrtateur de Quiz'; 
    }

    function desactiverQuiz(Quiz $quizz) {
        $quizz->estActive = false;
        echo "L'utilisateur $quizz->titre a été désactivé.";
    }

    function creerQuiz($titre, $typeDeQuiz, $cycleDeVie, $choix) {
        return new Quiz($titre, $typeDeQuiz, $cycleDeVie, $choix);
    }
}

class Createur extends Utilisateur {

    function __construct($nom, $prenom, $email, $password) {
        parent::__construct($nom, $prenom, $email, $password);
        $this->role = 'createur'; 
    }

    function creerQuiz($titre, $typeDeQuiz, $cycleDeVie, $choix) {
        return new Quiz($titre, $typeDeQuiz, $cycleDeVie, $choix);
    }
}

class UtilisateurParDefaut extends Utilisateur {

    function __construct($nom, $prenom, $email, $password) {
        parent::__construct($nom, $prenom, $email, $password);
        $this->role = 'Utilisateur'; 
    }

    function changerMotDePasse($newPassword, Utilisateur $user) {
        $user->password = $newPassword;
        echo "Mot de passe changé avec succès.";
    }
}

?>